std/experimental/allocator/building_blocks/allocator_list.d

```diff
@@ -632,119 +637,236 @@ version(Posix) @system unittest
     assert(b1.length == 1024 * 10);
 }
 
+/// Ditto
+shared struct SharedAllocatorList(Factory, BookkeepingAllocator = GCAllocator)
+{
+    import std.typecons : Ternary;
+    import core.internal.spinlock : SpinLock;
+    import std.traits : hasMember, isCallable, Parameters;
+    import std.algorithm.searching : startsWith;
+
+private:
+    // Forward all calls to 'impl' and protect them by the lock below
+    AllocatorList!(Factory, BookkeepingAllocator) impl;
+    SpinLock lock = SpinLock(SpinLock.Contention.brief);
+
+    // Remove this when 'https://github.com/dlang/druntime/pull/2156' gets merged
+    static ref T assumeUnshared(T)(ref shared T val) @trusted @nogc pure nothrow
+    {
+        return *cast(T*) &val;
+    }
+
+// Everything is inherited from the 'AllocatorList' implementation
+public:
+    /**
+    The alignment offered.
+    */
+    enum alignment = impl.alignment;
+
+    static foreach (T; __traits(allMembers, typeof(impl)))
+    {
+        static if (mixin("isCallable!(impl." ~ T ~ ")") && !T.startsWith("__"))
+        static if (mixin("__traits(getProtection, impl." ~ T ~ ") == `public`"))
+        {
+            mixin(`auto ` ~ T ~ `(Parameters!(impl.`~T~`) params)
+            {
+                lock.lock();
+                scope(exit) lock.unlock();
+                return assumeUnshared(impl).`~T~`(params);
+            }
+            `);
+        }
+    }
+}
```

Comment:
Note: this does not work well with rvalues because it copies them once more. Probably not a problem here because all parameters are cheap.

Reply:
I’ve added this note as a comment. See: d701391dfd725df4d1f0cec744ce3ee972cc0378
If it really affects someone they can open an issue or send a patch that changes it, I suppose.